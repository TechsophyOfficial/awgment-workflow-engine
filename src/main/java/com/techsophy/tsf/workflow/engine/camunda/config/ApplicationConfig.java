package com.techsophy.tsf.workflow.engine.camunda.config;

import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.servers.Server;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.converter.BufferedImageHttpMessageConverter;
import org.springframework.http.converter.HttpMessageConverter;
import java.awt.image.BufferedImage;
import java.util.List;
import static com.techsophy.tsf.workflow.engine.camunda.constants.CamundaRuntimeConstants.*;

@Configuration
public class ApplicationConfig
{
    @Bean
    public HttpMessageConverter<BufferedImage> httpMessageConverter() {
        return new BufferedImageHttpMessageConverter();
    }

    @Value(GATEWAY_URL)
    String gatewayUrl;
    @Bean
    public OpenAPI customOpenAPI()
    {
        return new OpenAPI()
                .info(new io.swagger.v3.oas.models.info.Info().title(WORKFLOW_ENGINE).version(VERSION_V1).description(WORKFLOW_ENGINE_API_VERSION_V1))
                .servers( List.of(new Server().url(gatewayUrl)));
    }
}
