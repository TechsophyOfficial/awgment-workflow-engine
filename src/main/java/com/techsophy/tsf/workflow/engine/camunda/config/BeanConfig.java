package com.techsophy.tsf.workflow.engine.camunda.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestTemplate;

/**
 * used to initialize beans
 */
@Configuration
public class BeanConfig
{
    /**
     * Bean initialization for rest template
     * @return
     */
    @Bean
    public RestTemplate restTemplate()
    {
        return new RestTemplate();
    }
}

