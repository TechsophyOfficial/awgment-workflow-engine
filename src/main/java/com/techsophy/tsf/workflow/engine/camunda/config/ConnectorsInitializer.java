package com.techsophy.tsf.workflow.engine.camunda.config;

import lombok.AllArgsConstructor;
import org.camunda.connect.Connectors;
import org.camunda.connect.spi.Connector;
import org.springframework.stereotype.Component;
import javax.annotation.PostConstruct;
import java.util.List;

/**
 * used to initialize connectors in camunda
 */
@Component
@AllArgsConstructor
public class ConnectorsInitializer extends Connectors
{
    private final List<Connector<?>> connectors;

    @PostConstruct
    public void initialize()
    {
        this.registerSpringDiscoveredConnectors();
    }

    /**
     * register connectors
     */
    private void registerSpringDiscoveredConnectors()
    {
        this.connectors.forEach(Connectors::registerConnector);
    }
}
