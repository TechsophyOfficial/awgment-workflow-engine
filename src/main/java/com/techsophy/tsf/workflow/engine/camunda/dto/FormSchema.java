package com.techsophy.tsf.workflow.engine.camunda.dto;

import lombok.Value;
import lombok.With;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Map;

/**
 * Form schema for form io
 */
@With
@Value
public class FormSchema
{
    @NotBlank(message = "name should not be null")
    private final String name;
    @NotBlank(message = "id should not be null")
    private final String id;
    @NotNull(message = "components should not be null")
    private final Map<String, Object> components;
    private final Integer version;
}
