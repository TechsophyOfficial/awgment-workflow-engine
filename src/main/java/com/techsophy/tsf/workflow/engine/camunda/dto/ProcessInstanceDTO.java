package com.techsophy.tsf.workflow.engine.camunda.dto;

import lombok.Value;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Map;

/**
 * Process instance DTO
 */
@Value
public class ProcessInstanceDTO
{
    @NotBlank(message = "processDefinitionKey should not be null")
    private final String processDefinitionKey;
    @NotNull(message = "variables should not be null")
    private final Map<String,Object> variables;
    @NotBlank(message = "businessKey should not be null")
    private final String businessKey;
    private final String formKey;
}
