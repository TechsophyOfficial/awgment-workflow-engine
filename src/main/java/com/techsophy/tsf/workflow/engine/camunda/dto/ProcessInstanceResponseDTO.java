package com.techsophy.tsf.workflow.engine.camunda.dto;

import lombok.AllArgsConstructor;
import lombok.Value;
import lombok.With;

import java.util.Map;

/**
 * Process instance response DTO
 */
@With
@Value
@AllArgsConstructor
public class ProcessInstanceResponseDTO
{
    private final  String processInstanceId;
    private final String businessKey;
    private final Map<String, Object> variables;
    private Boolean alreadyExists;
}
