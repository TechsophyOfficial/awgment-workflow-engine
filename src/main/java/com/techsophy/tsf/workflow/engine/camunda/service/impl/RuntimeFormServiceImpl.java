package com.techsophy.tsf.workflow.engine.camunda.service.impl;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.techsophy.tsf.workflow.engine.camunda.dto.FormSchema;
import com.techsophy.tsf.workflow.engine.camunda.model.ApiResponse;
import com.techsophy.tsf.workflow.engine.camunda.service.RuntimeFormService;
import com.techsophy.tsf.workflow.engine.camunda.service.TokenSupplier;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;
import java.util.Map;

import static com.techsophy.tsf.workflow.engine.camunda.constants.CamundaRuntimeConstants.*;

@Slf4j
@RefreshScope
@Service
@RequiredArgsConstructor
public class RuntimeFormServiceImpl implements RuntimeFormService
{
    private final TokenSupplier tokenSupplier;
    private final ObjectMapper objectMapper;
    private final RestTemplate restTemplate;
    private final WrapperServiceImpl wrapperService;
    private final WebClientWrapper webClientWrapper;
    @Value("${gateway.uri}")
    private String gatewayURI;

    @Override
    @Transactional
    public FormSchema fetchFormById(String formKey,String task)
    {
        HttpHeaders httpHeaders = new HttpHeaders();
        String url;
        String formVersion=LATEST;
        if(task!=null) {
            formVersion = wrapperService.getUserTaskExtensionByName(task, FORM_VERSION);
        }
        log.info("form version {}",formVersion);
        if(formVersion.equals(LATEST)){
            url=gatewayURI + FORM_RUNTIME_ENDPOINT_URL + "/" + formKey;
        }else{
            url=gatewayURI+FORM_MODELER_ENDPOINT_URL+formKey+"/"+formVersion;
        }
        httpHeaders.setAccept(List.of(MediaType.APPLICATION_JSON));
        httpHeaders.setContentType(MediaType.APPLICATION_JSON);
        httpHeaders.add(HttpHeaders.AUTHORIZATION, "Bearer "+tokenSupplier.getToken());
        HttpEntity<?> httpEntity = new HttpEntity<>(httpHeaders);
        ResponseEntity<?> responseEntity =  restTemplate.exchange(url, HttpMethod.GET, httpEntity, Map.class);
        ApiResponse<?> apiResponse = this.objectMapper.convertValue(responseEntity.getBody(), ApiResponse.class);
        return this.objectMapper.convertValue(apiResponse.getData(), FormSchema.class);
    }

    @SneakyThrows
    public FormSchema fetchFormById(String formKey)
    {
        String url=gatewayURI + FORM_RUNTIME_ENDPOINT_URL + "/" + formKey;
        WebClient webClient = webClientWrapper.createWebClient(tokenSupplier.getToken());
        String response = webClientWrapper.webclientRequest(webClient, url, HttpMethod.GET, String.class);
        ApiResponse<?> apiResponse = this.objectMapper.readValue(response, ApiResponse.class);
        return this.objectMapper.convertValue(apiResponse.getData(), FormSchema.class);
    }
}
