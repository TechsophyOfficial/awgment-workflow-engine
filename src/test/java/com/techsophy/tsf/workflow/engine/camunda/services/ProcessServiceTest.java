package com.techsophy.tsf.workflow.engine.camunda.services;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder;
import com.techsophy.tsf.workflow.engine.camunda.config.TestSecurityConfig;
import com.techsophy.tsf.workflow.engine.camunda.dto.FormSchema;
import com.techsophy.tsf.workflow.engine.camunda.dto.ProcessInstanceDTO;
import com.techsophy.tsf.workflow.engine.camunda.dto.ProcessInstanceResponseDTO;
import com.techsophy.tsf.workflow.engine.camunda.service.impl.RuntimeFormServiceImpl;
import com.techsophy.tsf.workflow.engine.camunda.service.impl.RuntimeProcessServiceImpl;
import lombok.AllArgsConstructor;
import lombok.Cleanup;
import org.camunda.bpm.engine.test.Deployment;
import org.junit.jupiter.api.*;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.get;
import static org.assertj.core.api.Assertions.assertThat;

@ActiveProfiles("test")
@ExtendWith(SnapshotExtension.class)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@AllArgsConstructor(onConstructor_ = {@Autowired})
@SpringBootTest(classes = TestSecurityConfig.class)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class ProcessServiceTest
{
    private static final String PROCESS_REQUEST_DATA = "testdata/process-request.json";
    private static final String CONTACT_DETAILS_DATA = "testdata/contact-details.json";
    private static final String CUSTOMER_DETAILS_DATA = "testdata/business-details.json";
    private static final String COMPLETE_TASK_DATA = "testdata/complete-task.json";

    private final ObjectMapper objectMapper;
    private final RuntimeFormServiceImpl formServiceTest;
    private final RuntimeProcessServiceImpl runtimeProcessService;
    private static final WireMockServer GATEWAY_URL=new WireMockServer(6666);

    @BeforeAll
    void init() throws IOException {
        GATEWAY_URL.start();
        @Cleanup InputStream contactDetails = new ClassPathResource(CONTACT_DETAILS_DATA).getInputStream();
        FormSchema form = this.objectMapper.readValue(contactDetails, FormSchema.class);
        GATEWAY_URL.stubFor(get("/form-runtime/v1/forms/contactDetails")
                .willReturn(json(objectMapper.writeValueAsString(Map.of(
                        "data",form,
                        "success",true,
                        "message","Form details detched successfully")))));
        @Cleanup InputStream businessDetails = new ClassPathResource(CUSTOMER_DETAILS_DATA).getInputStream();
        FormSchema businsessForm = this.objectMapper.readValue(businessDetails, FormSchema.class);
        GATEWAY_URL.stubFor(get("/form-runtime/v1/forms/businessDetails")
                .willReturn(json(objectMapper.writeValueAsString(Map.of(
                        "data",businsessForm,
                        "success",true,
                        "message","Form details detched successfully")))));
    }

    @AfterAll
    void destroy(){
        GATEWAY_URL.resetAll();
        GATEWAY_URL.stop();
    }

    public static ResponseDefinitionBuilder json(String apiResponse)
    {
        return aResponse().withHeader("Content-Type", MediaType.APPLICATION_JSON_VALUE).withBody(apiResponse);
    }



    @Test
    @Order(1)
    @Deployment(resources = {"testdata/api-controller.bpmn"})
    void testStartProcess() throws Exception
    {
        @Cleanup InputStream stream = new ClassPathResource(PROCESS_REQUEST_DATA).getInputStream();
        ProcessInstanceDTO process = this.objectMapper.readValue(stream, ProcessInstanceDTO.class);
        ProcessInstanceResponseDTO processResponse = runtimeProcessService.createProcessInstance(process);

        assertThat(processResponse.getBusinessKey()).isEqualTo(process.getBusinessKey());
    }

//    @Test
//    @Order(2)
//    void testNextTask()
//    {
//        FormSchemaResponse form = runtimeProcessService.getNextTask(BUSINESS_KEY);
//
//        expect(form.getFormContent(), form.getVariables(), form.getTaskName(), form.getFormKey()).orderedJson().toMatchSnapshot();
//    }
//
//    @Test
//    @Order(2)
//    void testCompleteTask() throws Exception
//    {
//        @Cleanup InputStream stream = new ClassPathResource(COMPLETE_TASK_DATA).getInputStream();
//        TaskDTO task = this.objectMapper.readValue(stream, TaskDTO.class);
//        FormSchemaResponse form = runtimeProcessService.completeAndFetchNextTask(task);
//
//        expect(form.getFormContent(), form.getVariables(), form.getTaskName(), form.getFormKey()).orderedJson().toMatchSnapshot();
//    }

//    @Test
//    @Order(3)
//    void testCompleteLastTask() throws Exception
//    {
//        @Cleanup InputStream stream = new ClassPathResource(COMPLETE_TASK_DATA).getInputStream();
//        TaskDTO task = this.objectMapper.readValue(stream, TaskDTO.class);
//        FormSchemaResponse form = runtimeProcessService.completeAndFetchNextTask(task);
//
//        assertThat(form).isNull();
//    }



}
