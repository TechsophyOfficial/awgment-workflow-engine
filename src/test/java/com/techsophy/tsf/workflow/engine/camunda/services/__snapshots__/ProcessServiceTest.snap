com.techsophy.tsf.workflow.engine.camunda.services.ProcessServiceTest.testCompleteTask=[
  {
    "components": [
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "myBusinessNeedLoanOf",
        "label": "My Business need loan of ",
        "mask": false,
        "multiple": false,
        "placeholder": "Loan Requirement",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "enterYourName",
        "label": "Enter Your Name",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "number",
        "label": "Number",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "number",
        "validate": {
          "customMessage": "",
          "max": 10,
          "min": 10,
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "email",
        "key": "email",
        "label": "Email",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "email",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "dateOfBirthAsPerPan",
        "label": "Date Of Birth (As per PAN)",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "datetime",
        "validate": {
          "customMessage": "",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "pan",
        "label": "PAN",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "maxLength": 10,
          "minLength": 10,
          "pattern": "[A-Z]{5}[0-9]{4}[A-Z]{1}",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "pincode",
        "label": "Pincode",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "maxLength": 6,
          "minLength": 6,
          "pattern": "[0-9]{6}|[0-9]{3}\\s[0-9]{3}",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "false",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "checkbox",
        "key": "receiveLoanRelatedUpdateOnWhatsApp",
        "label": "Receive Loan related update on WhatsApp",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "checkbox",
        "validate": {
          "customMessage": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": false,
        "key": "html",
        "label": "HTML",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "htmlelement",
        "validate": {
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": true,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "submit",
        "label": "Submit",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "button",
        "validate": {
          "required": false
        },
        "validateOn": "change"
      }
    ],
    "display": "form"
  },
  {
    "age": 25,
    "firstName": "alex",
    "lastname": "carey"
  },
  "Business Details",
  "businessDetails"
]


com.techsophy.tsf.workflow.engine.camunda.services.ProcessServiceTest.testNextTask=[
  {
    "components": [
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "myBusinessNeedLoanOf",
        "label": "My Business need loan of ",
        "mask": false,
        "multiple": false,
        "placeholder": "Loan Requirement",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "enterYourName",
        "label": "Enter Your Name",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "number",
        "label": "Number",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "number",
        "validate": {
          "customMessage": "",
          "max": 10,
          "min": 10,
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "email",
        "key": "email",
        "label": "Email",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "email",
        "validate": {
          "customMessage": "",
          "pattern": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "dateOfBirthAsPerPan",
        "label": "Date Of Birth (As per PAN)",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "datetime",
        "validate": {
          "customMessage": "",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "pan",
        "label": "PAN",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "maxLength": 10,
          "minLength": 10,
          "pattern": "[A-Z]{5}[0-9]{4}[A-Z]{1}",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "text",
        "key": "pincode",
        "label": "Pincode",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "textfield",
        "validate": {
          "customMessage": "",
          "maxLength": 6,
          "minLength": 6,
          "pattern": "[0-9]{6}|[0-9]{3}\\s[0-9]{3}",
          "required": true
        },
        "validateOn": "change"
      },
      {
        "defaultValue": "false",
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "inputType": "checkbox",
        "key": "receiveLoanRelatedUpdateOnWhatsApp",
        "label": "Receive Loan related update on WhatsApp",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "checkbox",
        "validate": {
          "customMessage": "",
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": false,
        "disabled": false,
        "hideLabel": false,
        "input": false,
        "key": "html",
        "label": "HTML",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "htmlelement",
        "validate": {
          "required": false
        },
        "validateOn": "change"
      },
      {
        "description": "",
        "disableOnInvalid": true,
        "disabled": false,
        "hideLabel": false,
        "input": true,
        "key": "submit",
        "label": "Submit",
        "mask": false,
        "multiple": false,
        "placeholder": "",
        "type": "button",
        "validate": {
          "required": false
        },
        "validateOn": "change"
      }
    ],
    "display": "form"
  },
  {
    "age": 25,
    "firstName": "alex",
    "lastname": "carey"
  },
  "Business Details",
  "businessDetails"
]